### PR TITLE
mikutter 3.4.6 (new formula)

### PR DESCRIPTION
**Pull request by midchildan**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

---

mikutter is an extensible twitter client. It is also available in the official repos of Debian, Gentoo, and BSDs.

### CODE REVIEW

File: Formula/mikutter.rb [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+
+  def install
+    mkpath libexec/"vendor"
+    resources.each do |r|
+      r.verify_download_integrity(r.fetch)
+      system("gem", "install", r.cached_download, "--no-document",
+             "--install-dir", "#{libexec}/vendor")
+    end
+
+    prefix.install_metafiles
```
Comment (dunn) [inline Formula/mikutter.rb]: If you're not changing the working directory during installation, this happens automatically.

File: Formula/mikutter.rb [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    rm_rf "vendor"
```
Comment (dunn) [inline Formula/mikutter.rb]: This shouldn't be necessary.

File: Formula/mikutter.rb [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    libexec.install Dir["*"]
+
+    if build.with? "terminal-notifier"
+      mkpath libexec/"plugin"
```
Comment (dunn) [inline Formula/mikutter.rb]: `#write` will create the path if necessary, so this and the rest of the `mkpath`s aren't needed: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/pathname.rb#L141

File: Formula/mikutter.rb [hunk 4]
```diff
@@ -0,0 +1,378 @@
+class Mikutter < Formula
+  desc "Extensible Twitter client."
+  homepage "http://mikutter.hachune.net"
+  url "http://mikutter.hachune.net/bin/mikutter.3.4.6.tar.gz"
+  sha256 "143318ad39ec60c9cf83e206f1a940e9464f0fda19c61e5e13a6cf327d1db076"
+  head "git://toshia.dip.jp/mikutter.git", :branch => "develop"
+
+  depends_on "gtk+"
+  depends_on "terminal-notifier" => :recommended
+  depends_on :ruby => "2.1+"
```
Comment (dunn) [inline Formula/mikutter.rb]: The RubyRequirement doesn't use the `+`, it should just be `"2.1"` and that means 2.1 or higher.

File: Formula/mikutter.rb [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    mkpath prefix/"mikutter.app/Contents/Resources"
+    system("/usr/bin/sips", "-s", "format", "icns",
+           libexec/"core/skin/data/icon.png",
+           "--out", "#{prefix}/mikutter.app/Contents/Resources/mikutter.icns")
+
+    (prefix/"mikutter.app/Contents/Info.plist").write(info_plist)
+  end
+
+  def exec_script
+    <<-EOS.undent
```
Comment (bfontaine) [inline Formula/mikutter.rb]: You should submit them to upstream instead.

Comment (MikeMcQuaid) [inline Formula/mikutter.rb]: > You should submit them to upstream instead.

Agreed. Upstream need to be willing to support macOS.

File: Formula/mikutter.rb (hunk 1 above)
Comment (midchildan) [inline Formula/mikutter.rb]: You mean `prefix.install_metafiles` ?

Comment (bfontaine) [inline Formula/mikutter.rb]: @midchildan Yes.

**Comment by MikeMcQuaid** [comment]
As a general comment: this formula is kinda a full port to macOS. I'd strongly recommend workign with upstream to get all this stuff (including e.g. icns generation) upstream and ideally get a .app bundle that they can build and support.

File: Formula/mikutter.rb [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+          if user
+            user_img = Gdk::WebImageLoader.local_path(user[:profile_image_url])
+            metadata[:title] = "@\#{user[:idname]} (\#{user[:name]})"
+            metadata[:contentImage] = user_img
+          end
+          text = text.to_s if text.is_a? Message
+          metadata[:message] = text
+
+          command = [
+            "#{HOMEBREW_PREFIX}/opt/terminal-notifier/bin/terminal-notifier",
```
Comment (bfontaine) [inline Formula/mikutter.rb]: You can’t rely on this being present because it’s only `:recommended`; not required.

File: Formula/mikutter.rb (hunk 5 above)
Comment (bfontaine) [inline Formula/mikutter.rb]: You don’t have to depend on `HOMEBREW_PREFIX`; just use the binaries available in the `PATH`.

File: Formula/mikutter.rb (hunk 6 above)
Comment (bfontaine) [inline Formula/mikutter.rb]: Even if you could you should use `terminal-notifier` instead of its full path.

**Comment by midchildan** [comment]
@MikeMcQuaid The compatibility plugin in this formula could easily be provided by upstream or elsewhere, but it is not so easy for the rest. First, mikutter depends on formulas provided by Homebrew, and you will have to include every single one of them to package it as a standalone `.app` bundle. I'm not sure the upstream will be willing to do that. Second, mikutter depends on Hombrew-specific info such as the `HOMEBREW_PREFIX`. This is necessary because apps launched from the Finder will not include Homebrew's `PATH` by default. 
But the approach I used to make an `.app` bundle is applicable to other formulas as well.  Most of the wrapper script I wrote for launching mikutter is about dealing with differences in the way macOS and other common unix systems handle locales. Specifically, many unix apps depends on `LANG` being set to a UTF-8 locale and will not work if `LANG` is unspecified. Can't we perhaps add a helper method to make a `.app` bundle to wrap unix GUI apps?

File: Formula/mikutter.rb (hunk 5 above)
Comment (midchildan) [inline Formula/mikutter.rb]: I could do that, but I would still need `HOMEBREW_PREFIX` for reasons given below.

**Comment by midchildan** [comment]
I opened a PR in Homebrew/brew#1383

**Comment by midchildan** [comment]
Fixed formula.

File: Formula/mikutter.rb [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+  head "git://toshia.dip.jp/mikutter.git", :branch => "develop"
+
+  depends_on "gtk+"
+  depends_on "terminal-notifier" => :recommended
+  depends_on :ruby => "2.1"
+
+  resource "rake" do
+    url "https://rubygems.org/gems/rake-10.5.0.gem"
+    sha256 "2b55a1ad44b5c945719d8a97c302a316af770b835187d12143e83069df5a8a49"
+  end
```
Comment (MikeMcQuaid) [inline Formula/mikutter.rb]: Can you add a blank line between each of these? Thanks!

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

File: Formula/mikutter.rb (hunk 7 above)
Comment (midchildan) [inline Formula/mikutter.rb]: @MikeMcQuaid done😁

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks for your first contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**